apps/glusterfs/allocator_simple.go

```diff
@@ -89,15 +95,15 @@ func (s *SimpleAllocator) AddDevice(cluster *ClusterEntry,
 	node *NodeEntry,
 	device *DeviceEntry) error {
 
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
 	// Create a new cluster id if one is not available
 	clusterId := cluster.Info.Id
-	if _, ok := s.rings[clusterId]; !ok {
-		s.rings[clusterId] = NewSimpleAllocatorRing()
+	// TODO: in the future, we should do this call separately
+	if err := s.AddCluster(clusterId); err != nil {
```

Comment:
Is this correct? This will bail out if the cluster already exists...

Reply:
Unfortunately, that *is* correct. AFAICT. This is technical debt speaking. Someone made an early mistake to have a "cluster" added silently if we added a "device", but the cluster was missing. They should have been two separate pieces but they are intertwined everywhere. Eventually in a down-the-road-part-two we'll fix this up if we get there, which is why I explicitly added the TODO there, and handle the error.

You'll note the AddCluster currently doesn't ever error, but some future patch could change that, hence handling the error correctly now.

Reply:
Side rant: How on earth can I properly reply in this horrbile github thing? Do I always need to copy and paste the previous comment myself... :-(


> Unfortunately, that is correct. AFAICT. This is technical debt speaking. Someone made an early mistake to have a "cluster" added silently if we added a "device", but the cluster was missing. They should have been two separate pieces but they are intertwined everywhere. Eventually in a down-the-road-part-two we'll fix this up if we get there, which is why I explicitly added the TODO there, and handle the error.

I hear what you are saying about how it can be improved. But you are changing this behavior here: You error out in `AddDevice` if the cluster already exist. (Because your `AddCluster` errors out in that case.) Hence I think this is not correct. If it is, I have not understood your explanation above, and please be more specific.

> You'll note the AddCluster currently doesn't ever error, but some future patch could change that, hence handling the error correctly now.

AFAICT this is not correct: As I said in my comment AddCluster errors out if the cluster already exists. Quoting from your patch (that adds AddCluster):

```
+	if _, ok := s.rings[clusterId]; ok {
+		logger.LogError("cluster id %s already exists", clusterId)
+		return ErrFound
+	}
```

How is this not erroring out?

Reply:
You're not looking at the correct version of the patch. I suspect you're getting confused by the GitHub UI. That diff doesn't exist in the latest patch.

Reply:
No, I am looking exactly at the current version of the patch. Have you pushed your latest changes? ...

Reply:
yup, see here for a direct link: https://github.com/heketi/heketi/pull/850/files/7eae99db3e15bd77887050eaf79fa26c4e60876f#diff-36d2e1cdec2dabf576d0bc9e6deda2fcR148